ethcore/src/evm/wasm/runtime.rs

```diff
@@ -0,0 +1,356 @@
+
+		Ok(None)
+	}
+
+	/// Invoke create in the state runtime
+	pub fn create(&mut self, context: interpreter::CallerContext) 
+		-> Result<Option<interpreter::RuntimeValue>, interpreter::Error>
+	{
+		//
+		// method signature:
+		//   fn create(endowment: *const u8, code_ptr: *const u8, code_len: u32, result_ptr: *mut u8) -> i32;
+		//
+
+		trace!(target: "wasm", "runtime: create contract");
+		let mut context = context;
+		let result_ptr = context.value_stack.pop_as::<i32>()? as u32;
+		trace!(target: "wasm", "    result_ptr: {:?}", result_ptr);		
+		let code_len = context.value_stack.pop_as::<i32>()? as u32;
+		trace!(target: "wasm", "    code_len: {:?}", code_len);		
+		let code_ptr = context.value_stack.pop_as::<i32>()? as u32;
+		trace!(target: "wasm", "    code_ptr: {:?}", code_ptr);		
+		let endowment = self.pop_u256(&mut context)?;
+		trace!(target: "wasm", "    val: {:?}", endowment);
+
+		let code = self.memory.get(code_ptr, code_len as usize)?;
+
+		let gas_left = self.gas_left()
+			.map_err(|_| interpreter::Error::Trap("Gas state error".to_owned()))?
+			.into();
+		
+		match self.ext.create(&gas_left, &endowment, &code, evm::CreateContractAddress::FromSenderAndCodeHash) {
+			evm::ContractCreateResult::Created(address, gas_left) => {
+				self.memory.set(result_ptr, &*address)?;
+				self.gas_counter = self.gas_limit - gas_left.low_u64();
+				trace!(target: "wasm", "runtime: create contract success (@{:?})", address);
+				Ok(Some(0i32.into()))
+			},
+			evm::ContractCreateResult::Failed => {
+				trace!(target: "wasm", "runtime: create contract fail");
+				Ok(Some((-1i32).into()))
+			}
+		}
+	}
+
+	/// Allocate memory using the wasm stack params
+	pub fn malloc(&mut self, context: interpreter::CallerContext) 
+		-> Result<Option<interpreter::RuntimeValue>, interpreter::Error>
+	{
+		let amount = context.value_stack.pop_as::<i32>()? as u32;
+		let previous_top = self.dynamic_top;
+		self.dynamic_top = previous_top + amount;
+		Ok(Some((previous_top as i32).into()))
+	}
+
+	/// Allocate memory in wasm memory instance
+	pub fn alloc(&mut self, amount: u32) -> Result<u32, Error> {
+		let previous_top = self.dynamic_top;
+		self.dynamic_top = previous_top + amount;
+		Ok(previous_top.into())
+	}
+
+	/// Report gas cost with the params passed in wasm stack
+	fn gas(&mut self, context: interpreter::CallerContext) 
+		-> Result<Option<interpreter::RuntimeValue>, interpreter::Error> 
+	{
+		let amount = context.value_stack.pop_as::<i32>()? as u64;
+		if self.charge_gas(amount) {
+			Ok(None)
+		} else {
+			Err(interpreter::Error::Trap(format!("Gas exceeds limits of {}", self.gas_limit)))			
+		}
+	}
+
+	fn charge_gas(&mut self, amount: u64) -> bool {
+		let prev = self.gas_counter;
+		if prev + amount > self.gas_limit {
+			// exceeds gas
+			false
+		} else {
+			self.gas_counter = prev + amount;
+			true
+		}		
+	}
+
+	fn h256_at(&self, ptr: WasmPtr) -> Result<H256, interpreter::Error> {
```

Comment:
Bound-checked, sure, but that doesn't mean that it won't cause type confusion. We're for all intents and purposes doing unchecked `mem::transmute` to synthesise these values. It's not going to access uninitialized memory but it may cause us to interpret data as the wrong type.